### PR TITLE
Add proposal to adopt C++11 as the base language

### PR DESCRIPTION
**Pull request by llvm-beanz**
This proposal is for HLSL 202y to adopt C++ 11 as its base C++ language version and encorporate as many of C++ 11's features as reasonably fit within HLSL's language design.

### CODE REVIEW

File: proposals/NNNN-cxx11-base.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+* decltype
+* constexpr
+* C++11 scoped enumerations
+* variadic templates
+* user-defined literals
+* [C++11 attributes](/proposals/0002-cxx-attributes.md)
+* Lambda expressions
+* Static assert
+* Range-based for loops
+
```
Comment (bharadwajy) [inline proposals/NNNN-cxx11-base.md:36]: The above listed set of core C++11 language features appear to be a strict subset of the [full set](https://en.cppreference.com/w/cpp/11). Is the proposal to support only this subset for now with an intent to incrementally support others as deemed necessary?

Comment (llvm-beanz) [reply proposals/NNNN-cxx11-base.md:36]: Yes, in general I think that we should never wholesale adopt all changes in a C++ version. I think we should identify specific features we want and don't want and consciously think through how those features interact with existing HLSL language features.

Each of the features I listed here needs some specific design considerations, but IMO, they're all strong candidates for inclusion in HLSL 202y.

**Review by bharadwajy — APPROVED** [review]

File: proposals/NNNN-cxx11-base.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+base C++ language to C++ 2011.
+
+## Motivation
+
+C++11 is over a decade old and introduced widely adopted features, many of which
+have been frequently requested additions for HLSL.
+
+## Proposed solution
+
+Adopt a C++ 11 base language and include the following C++11 features in HLSL:
```
Comment (damyanp) [inline proposals/NNNN-cxx11-base.md:25]:
```suggestion
Adopt a C++ 11 base language and include the following C++11 features in HLSL 202y:

```

Comment (damyanp) [inline proposals/NNNN-cxx11-base.md:25]:
```suggestion
Adopt a C++ 11 base language and include the following C++ 11 features in HLSL:

```

File: proposals/NNNN-cxx11-base.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+* Sponsor: TBD
+* Status: **Under Consideration**
+* Planned Version: 202y
+
+## Introduction
+
+In DXC HLSL is a set of feature extensions on top of a subset of C++ 98. C++ 98
+is now over 20 years old and most modern C++ users have adopted newer language
+constructs. This proposal suggests taking the small step of updating HLSL 202y's
+base C++ language to C++ 2011.
```
Comment (damyanp) [inline proposals/NNNN-cxx11-base.md:16]:
```suggestion
base C++ language to C++ 11.

```

File: proposals/NNNN-cxx11-base.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+## Motivation
+
+C++11 is over a decade old and introduced widely adopted features, many of which
```
Comment (damyanp) [inline proposals/NNNN-cxx11-base.md:20]:
```suggestion
C++ 11 is over a decade old and introduced widely adopted features, many of which

```

**Review by damyanp — APPROVED** [review]
Sounds great to me!

Some nits in the comments around consistency between `C++ 2021` vs `C++11` vs `C++ 11`.  I started calling them out but realized that there's quite a few (including some `C++20`s and you might not necessarily agree with me on which way you want to standardize, if at all.

**Comment by llvm-beanz** [comment]
> Some nits in the comments around consistency between `C++ 2021` vs `C++11` vs `C++ 11`.

Consistency is good. I think the normalized spelling for C++ versions is "C++NN" (no space). Are you okay with me doing a pass to normalize that way?

**Comment by damyanp** [comment]
> Are you okay with me doing a pass to normalize that way?

Absolutely - I have no real preference other than that we're consistent!

**Comment by llvm-beanz** [comment]
Assigned this proposal number 23

**Pull request merged**